tests/core/test_variable.py

```diff
@@ -0,0 +1,558 @@
+        self.assertEqual(state, v.State.FREE)
+
+    def test_value(self):
+        """Test methods for setting values and checking bounds."""
+        #test construction with value between bounds
+        v = relentless.DesignVariable(value=0.0, low=-1.0, high=1.0)
+        self.assertAlmostEqual(v.value, 0.0)
+        self.assertEqual(v.state, v.State.FREE)
+
+        #test below low
+        v.value = -1.5
+        self.assertAlmostEqual(v.value, -1.0)
+        self.assertEqual(v.state, v.State.LOW)
+
+        #test above high
+        v.value = 3
+        self.assertAlmostEqual(v.value, 1.0)
+        self.assertEqual(v.state, v.State.HIGH)
+
+        #test invalid value
+        with self.assertRaises(ValueError):
+            v.value = '0'
+
+    def test_bounds(self):
+        """Test methods for setting bounds and checking value clamping."""
+        #test construction with value between bounds
+        v = relentless.DesignVariable(value=0.0, low=-1.0, high=1.0)
+        self.assertAlmostEqual(v.value, 0.0)
+        self.assertEqual(v.state, v.State.FREE)
+
+        #test changing low bound to above value
+        v.low = 0.2
+        self.assertAlmostEqual(v.value, 0.2)
+        self.assertAlmostEqual(v.low, 0.2)
+        self.assertAlmostEqual(v.high, 1.0)
+        self.assertEqual(v.state, v.State.LOW)
+
+        #test changing high bound to below value
+        v.low = -1.0
+        v.high = -0.2
+        self.assertAlmostEqual(v.value, -0.2)
+        self.assertAlmostEqual(v.low, -1.0)
+        self.assertAlmostEqual(v.high, -0.2)
+        self.assertEqual(v.state, v.State.HIGH)
+
+        #test changing bounds incorrectly
+        with self.assertRaises(ValueError):
+            v.low = 0.2
+        with self.assertRaises(ValueError):
+            v.high = -1.6
+
+class DepVar(relentless.DependentVariable):
+    """Mock dependent variable to test relentless.DependentVariable"""
+    def __init__(self, *a, **b):
+        super().__init__(*a, **b)
+
+    @property
+    def value(self):
+        return np.sum([v.value for p,v in self.depends])
+
+    def _derivative(self, param):
+        if param in self.params:
+            if getattr(self, param).value > 5.0:
+                return 0.5
+            else:
+                return -0.5
+        else:
+            raise ValueError('Unknown parameter')
+
+class test_DependentVariable(unittest.TestCase):
+    """Unit tests for relentless.DependentVariable"""
+
+    def test_init(self):
+        """Test creation with data."""
+        t = relentless.DesignVariable(value=1.0)
+        u = relentless.DesignVariable(value=2.0)
+        v = relentless.DesignVariable(value=3.0)
+
+        #test creation with only vardicts
+        w = DepVar({'t':t, 'u':u, 'v':v})
+        self.assertCountEqual(w.params, ('t','u','v'))
+        self.assertDictEqual({p:v for p,v in w.depends}, {'t':t,'u':u,'v':v})
+        self.assertAlmostEqual(w.value, 6.0)
+
+        #test creation with only kwvars
```

Comment:
I think this is missing the constructor for `w`?
```suggestion
        #test creation with only kwvars
        w = DepVar(t=t, u=u, v=v)
```